cpp/src/arrow/acero/swiss_join_avx2.cc

```diff
@@ -233,27 +243,288 @@ int RowArrayAccessor::VisitNulls_avx2(const RowTableImpl& rows, int column_id,
+  __m256i row_lo = _mm256_i64gather_epi64(row_ptr_base_i64, offset_lo, 1);
+  __m256i row_hi = _mm256_i64gather_epi64(row_ptr_base_i64, offset_hi, 1);
+  _mm256_storeu_si256(reinterpret_cast<__m256i*>(output), row_lo);
+  _mm256_storeu_si256(reinterpret_cast<__m256i*>(output + 4), row_hi);
+}
+
+inline void Decode1_avx2(uint8_t* output, const uint8_t* row_ptr, uint32_t num_bytes) {
+  // Copy 32 bytes at a time.
+  __m256i* output_i256 = reinterpret_cast<__m256i*>(output);
+  const __m256i* row_ptr_i256 = reinterpret_cast<const __m256i*>(row_ptr);
+  for (int istripe = 0; istripe < bit_util::CeilDiv(num_bytes, 32); ++istripe) {
+    _mm256_storeu_si256(output_i256 + istripe,
+                        _mm256_loadu_si256(row_ptr_i256 + istripe));
+  }
+}
+
+inline uint32_t Decode8Offset_avx2(uint32_t* output, uint32_t current_length,
+                                   __m256i num_bytes) {
+  uint32_t num_bytes_last = static_cast<uint32_t>(_mm256_extract_epi32(num_bytes, 7));
+  // Init every offset with the current length.
+  __m256i offsets = _mm256_set1_epi32(current_length);
+  // We keep right-shifting the length and accumulate the offset by adding the length.
+  __m256i length =
+      _mm256_permutevar8x32_epi32(num_bytes, _mm256_setr_epi32(7, 0, 1, 2, 3, 4, 5, 6));
+  length = _mm256_insert_epi32(length, 0, 0);
+  for (int i = 0; i < 7; ++i) {
+    offsets = _mm256_add_epi32(offsets, length);
+    length =
+        _mm256_permutevar8x32_epi32(length, _mm256_setr_epi32(7, 0, 1, 2, 3, 4, 5, 6));
+    length = _mm256_insert_epi32(length, 0, 0);
+  }
+  _mm256_storeu_si256(reinterpret_cast<__m256i*>(output), offsets);
+  return _mm256_extract_epi32(offsets, 7) + num_bytes_last;
+}
+
+inline void Decode8Null_avx2(uint8_t* output, uint64_t null_bytes) {
+  uint8_t null_bits =
+      static_cast<uint8_t>(_mm256_movemask_epi8(_mm256_set1_epi64x(null_bytes)));
+  *output = ~null_bits;
+}
+
+}  // namespace
+
+int RowArray::DecodeFixedLength_avx2(ResizableArrayData* output, int output_start_row,
+                                     int column_id, uint32_t fixed_length,
+                                     int num_rows_to_append,
+                                     const uint32_t* row_ids) const {
+  // Benchmarking shows that when the data element width is <= 8, the scalar code almost
+  // always outperforms the vectorized code - about 2X~3X faster when the whole data set
+  // falls into L1~LLC, and the ratio goes down to about 1:1 as the data size increases
+  // when most of the accesses hit the main memory. This is possibly because that decoding
+  // is mostly copying scattered pieces of data and there is not enough computation to pay
+  // off the cost of the heavy gather instructions.
+  // For fixed length 0 (boolean column), the vectorized code wins by batching 8 bits into
+  // a single byte instead of modifying the same byte 8 times in the scalar code.
```

Comment:
Given that it doesn't affect AMD processors nor recent Intel processors, we could simply keep the optimization enabled.

Reply:
OK, that's good for me too. I've removed the fixed-length-specific checking (and the comment).

I'm getting the final benchmarking numbers and will un-draft the PR afterwards.

Reply:
+1 for me, and this should be denote in Release note?

Reply:
Nice reminder!

@raulcd I think I can put something like "the performance improvement on specific CPU models (blablabla) may not be as significant as expected due to blablabla" in the PR description. Is there something we should do to ensure that will appear in the coming release notes?

Thanks.

Reply:
I've put something in the "Are there any user-facing changes?" section of the PR description. Quote:

> No changes other than positive performance improvement. Users can expect such improvement for hash joins related workload. Nevertheless the improvement degree highly depends on not only the workload, but also the CPU models. For Intel CPUs from Skylake to Icelake/Tigerlake, which suffer the performance degradation of AVX2 gather because of an vulnerability mitigation of Intel's (detailed in https://github.com/apache/arrow/pull/43832#issuecomment-2326646353), the improvement is less significant - single digit percent. Other models, e.g. AMD, and the most recent Intel, can achieve better improvement up to 30%.